lib/cc/analyzer/engine.rb

```diff
@@ -35,32 +34,30 @@ def run(stdout_io, stderr_io = StringIO.new)
           end
         end
 
-        Thread.new do
-          sleep TIMEOUT
-
-          if engine_running
-            Thread.current.abort_on_exception = true
-            run_command("docker kill #{container_name}")
-
-            Analyzer.statsd.increment("cli.engines.result.error")
-            Analyzer.statsd.increment("cli.engines.result.error.timeout")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error.timeout")
-            raise EngineTimeout, "engine #{name} ran past #{TIMEOUT} seconds and was killed"
+        begin
+          status = nil
+          Timeout::timeout(TIMEOUT) do
+            pid, status = Process.waitpid2(pid)
           end
+        rescue Timeout::Error
+          run_command("docker kill #{container_name} || true")
+          timed_out = true
```

Comment:
Calling `waitpid` again after sending the kill should reap the process.

On Sun, Jun 28, 2015, 19:15 Bryan Helmkamp <notifications@github.com> wrote:

> In lib/cc/analyzer/engine.rb
> <https://github.com/codeclimate/codeclimate/pull/35#discussion_r33428252>:
>
> >            end
> > +        rescue Timeout::Error
> > +          run_command("docker kill #{container_name} || true")
> > +          timed_out = true
>
> In the timeout case, the process does not get cleaned up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/codeclimate/codeclimate/pull/35/files#r33428252>.
>